### PR TITLE
🏗️ enable 2-at-a-time batching for all sauce labs tests 

### PR DESCRIPTION
**Pull request by rcebulko**
### What's going on

We have 15 concurrency slots in our Sauce Labs plan. This is not the number of open tunnels; it's the number of browser test sessions we have running at any one time. When we issue test requests in excess of the limit, Sauce Labs will try to queue them for a short time; according to [Sauce](https://support.saucelabs.com/hc/en-us/articles/225243308-Why-am-I-not-getting-the-parallelism-concurrency-I-expected-):

> Queuing is intended to deal with small mistakes in concurrency management. Sauce Labs does not recommend queuing tests intentionally.
>
> Tests that are queued for too long will time out with an error. Tests may also time out if too many tests are already queued.  Sauce Labs reserves the right to alter queue size, time and availability at will.

When we run `gulp integration --saucelabs --stable`, it triggers \~6 browser runs simultaneously. Some complete quickly (like IE11, which runs only one test) while others are much slower (ie. Chrome latest and canary, which run all tests). This full parallelism means that a single PR consumes a peak of 5\~6 concurrency sessions. During peak contribution hours (\~5pm EST == \~2pm PST), the high PR  volume means a high number of (both pull and push) Travis jobs running tests on Sauce.

As the queue gets long--our peak concurrency in the last month is usually around 25--some browser sessions never run, reporting timeout errors in Travis and failing the tests. This in turn causes the developer to re-start the Travis job, re-queue the Sauce sessions, and in turn delay/break other Travis jobs.

### What this does

This re-enables batching functionality for Sauce. We use it already for unit tests, but it was disabled for integration tests a while back because we didn't think it was needed. Instead of running 5~6 browsers at a time, it batches requests into at most 2 at a time. This will limit any given Travis job to creating 2 sessions, and should drastically reduce our peak session load.

By looking at 10 previous (successful) Sauce Labs jobs on Travis, I determined the mean runtime to be 1842s with a standard deviation of 137s. This yields a 95% confidence-interval of [1568s, 2117]. Test runs of these changes resulted in a runtime of 2038s, indicating there was not a statistically significant increase in the runtime due to the restricted concurrency.

### Why 2 at a time instead of 3 or 4

Why not? Increasing concurrency (or leaving it as unlimited) clearly didn't significantly reduce runtime, even when all tests run successfully. Reducing it to 2 is sufficient to maintain good performance, and has the potential to drastically reduce Sauce test flakiness by addressing the concurrency limit issue.

Could we increase the batch size to 3? Probably, for now. But a) it wouldn't actually reduce runtime (as evidence shows) b) it would increase the likelihood of Sauce flaking during high-activity periods and c) it will have to be adjusted further if/when the developer team grows in such a way as to increase our PR rate (or during events like Fixits where there is a surge).

### CODE REVIEW

**Comment by rsimha** [comment]
Couple high level comments before I review in greater detail:

1. This PR breaks test status reporting for the sauce labs tasks.
Screen Shot 2020-03-01 at 1 57 35 PM


2. The total running time of the sauce labs task has increased by 30%, from [~27 min](https://travis-ci.org/ampproject/amphtml/jobs/656515596#L2449) to [~35 min](https://travis-ci.org/ampproject/amphtml/jobs/656393373#L2486). This will noticeably increase wall-clock time for Travis jobs. Having said that, I do agree that going from unbatched runs to batched runs will improve stability today (even if they didn't seem necessary a few months ago). Can you try variants of this PR with batch sizes of 3 (and 4) for comparison? I suspect the sweet spot may be 3, but we'd need actual running times to verify.

**Comment by rcebulko** [comment]
> 1. This PR breaks test status reporting for the sauce labs tasks.

Good catch; I'll address that.

> 1. The total running time of the sauce labs task has increased by 30%, from [~27 min](https://travis-ci.org/ampproject/amphtml/jobs/656515596#L2449) to [~35 min](https://travis-ci.org/ampproject/amphtml/jobs/656393373#L2486). This will noticeably increase wall-clock time for Travis jobs. Having said that, I do agree that going from unbatched runs to batched runs will improve stability today (even if they didn't seem necessary a few months ago). Can you try variants of this PR with batch sizes of 3 (and 4) for comparison? I suspect the sweet spot may be 3, but we'd need actual running times to verify.

I'm a bit confused by what happened here. When I created this PR, the Travis task and logs listed a runtime of 33m 58s for this push build; it now shows 36m 59s. It's unclear to me why this is.

Regardless of the cause, given that the new value falls outside of the CI range, the data support that there is a statistically significant runtime increase. Early this week I'll spin up branches of this to experiment with different batch sizes.

**Comment by estherkim** [comment]
> The total running time of the sauce labs task has increased by 30%, from [~27 min](https://travis-ci.org/ampproject/amphtml/jobs/656515596#L2449) to [~35 min](https://travis-ci.org/ampproject/amphtml/jobs/656393373#L2486). 

The 27 min job is a master build. We should be comparing against pull request builds because Travis configs vary slightly between the two. This PR build, for example, goes for [31 min](https://travis-ci.org/ampproject/amphtml/jobs/657247451).

Do we have a goal # of parallel jobs we want to support? Might help in determining the sweet spot of `parallel jobs x batch size x build duration`, given the concurrency limit of 15.

**Comment by rcebulko** [comment]
> > The total running time of the sauce labs task has increased by 30%, from [~27 min](https://travis-ci.org/ampproject/amphtml/jobs/656515596#L2449) to [~35 min](https://travis-ci.org/ampproject/amphtml/jobs/656393373#L2486).
> 
> The 27 min job is a master build. We should be comparing against pull request builds because Travis configs vary slightly between the two. This PR build, for example, goes for [31 min](https://travis-ci.org/ampproject/amphtml/jobs/657247451).

This difference can be seen when compared against [this](https://travis-ci.org/ampproject/amphtml/jobs/656001520?utm_medium=notification&utm_source=github_status) and [this](https://travis-ci.org/ampproject/amphtml/jobs/656001571?utm_medium=notification&utm_source=github_status) PR builds, which were baselines I ran at the same test. I think the takeaway is that there is some small but not insignificant runtime increase.

> Do we have a goal # of parallel jobs we want to support? Might help in determining the sweet spot of `parallel jobs x batch size x build duration`, given the concurrency limit of 15.

It's tricky to start at # of Travis jobs and work up, given the irregular duration of Sauce sessions and durations each Travis job kicks out. There are so many moving parts at play here, I feel like the best we can do is adjust based on peak concurrency (and, eventually, upgrade to higher concurrency limits)

**Comment by rsimha** [comment]
Good discussion all around. Some more comments:

- I was surprised to learn that 15 "sessions" meant VMs and not tunnels. I think this fact alone makes a case for reintroducing a batch size for stable tests.
- Re: the fluctuation in job running times, it could be due to non-hermetic tests, or varying loads during different times of the day.
- The running times I cited did not include Travis VM boot-up time, and included only this number:
image Overall, I agree with @rcebulko's statement "_there is some small but not insignificant runtime increase_" 
- Re: @estherkim's observation "_Travis configs vary slightly between the two_",  I believe we run the same set of commands in the remote tests job for push and PR builds: download packages, start sauce connect, unit saucelabs, integration saucelabs stable, integration saucelabs beta, stop sauce connect. I'd imagine any substantial increase / decrease in running time will apply to both build types.

Re: next steps for this PR, perhaps we should reintroduce batching and pick the best batch size (2 or 3) based on what testing reveals as reasonable? WDYT?

**Comment by rcebulko** [comment]
> * Re: the fluctuation in job running times, it could be due to non-hermetic tests, or varying loads during different times of the day.

FWIW, I intentionally started the Sauce Labs test runs at the exact same time across this and the control branches, so load variations should be pretty uniform across.

> * The running times I cited did not include Travis VM boot-up time, and included only this number:
>   image Overall, I agree with @rcebulko's statement "_there is some small but not insignificant runtime increase_"

> * Re: @estherkim's observation "_Travis configs vary slightly between the two_",  I believe we run the same set of commands in the remote tests job for push and PR builds: download packages, start sauce connect, unit saucelabs, integration saucelabs stable, integration saucelabs beta, stop sauce connect. I'd imagine any substantial increase / decrease in running time will apply to both build types.

I was comparing against control-branch PR builds, not against master builds anyway, so I believe this is moot.

> Re: next steps for this PR, perhaps we should reintroduce batching and pick the best batch size (2 or 3) based on what testing reveals as reasonable? WDYT?

In https://github.com/ampproject/amphtml/pull/27051 and https://github.com/ampproject/amphtml/pull/27052 I've altered the batch sizes to 3 and 4 respectively. Right now all three are running; I'll update this thread with runtimes to compare once they are completed.

**Comment by estherkim** [comment]
> The running times I cited did not include Travis VM boot-up time, and included only this number:
image Overall, I agree with @rcebulko's statement "there is some small but not insignificant runtime increase"

We should take into account Travis build times, not the remote job test. There could be background processes, i.e. an increased # of connections to Sauce Labs, that can affect VM shut down. In the end, it's the Travis build time that affects developer productivity.

> * Re: @estherkim's observation "_Travis configs vary slightly between the two_",  I believe we run the same set of commands in the remote tests job for push and PR builds: download packages, start sauce connect, unit saucelabs, integration saucelabs stable, integration saucelabs beta, stop sauce connect. I'd imagine any substantial increase / decrease in running time will apply to both build types.

We don't store cache on master builds but agreed with your conclusion.

**Comment by rcebulko** [comment]
Update: I ran three sets of Sauce Labs tests: this PR with 2-at-a-time batching, [3-at-a-time](https://github.com/ampproject/amphtml/pull/27051), and [4-at-a-time](https://github.com/ampproject/amphtml/pull/27052). I ran all three at once in five separate rounds. I won't report all the stats here, but I ran one-tailed paired t-tests to compare 2/3 and 3/4. 2-at-a-time is significantly slower than 3-at-a-time (with 98.4% confidence), while there is no significant difference between batch sizes of 3 and 4.

TL;DR - @rsimha 's prediction was correct that 3 is indeed the sweet spot. I'm going to close this PR and link back to this thread in https://github.com/ampproject/amphtml/pull/27051
